### PR TITLE
executor: fix index join error when join key is ENUM or SET

### LINKED ISSUE #19233: Index join returns error when join key is enum/set
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
drop table if exists i;
drop table if exists p;
drop table if exists t;
CREATE TABLE `p` (
    `type` enum('HOST_PORT') NOT NULL,
    UNIQUE KEY (`type`)
) ;

CREATE TABLE `i` (
  `objectType` varchar(64) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

insert into i values ('SWITCH');

create table t like i;
insert into t values ('HOST_PORT');
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;

insert into i select * from t;

insert into p values('HOST_PORT');
select /*+ INL_HASH_JOIN(p) */ * from i, p where i.objectType = p.type;
```

### 2. What did you expect to see? (Required)

```mysql
select /*+ INL_HASH_JOIN(p) */ * from i, p where i.objectType = p.type;
...
65536 rows in set (0.05 sec)
```

### 3. What did you see instead (Required)

```mysql
select /*+ INL_HASH_JOIN(p) */ * from i, p where i.objectType = p.type;
ERROR 1265 (01000): Data truncated for column '%s' at row %d
```

And panic occurs in the log
```
[2020/08/17 15:47:32.075 +08:00] [ERROR] [misc.go:91] ["panic in the recoverable goroutine"] [r="\"invalid memory address or nil pointer dereference\""] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/util/misc.go:93\nruntime.gopanic\n\t/usr/local/Cellar/go@1.13/1.13.9/libexec/src/runtime/panic.go:679\nruntime.panicmem\n\t/usr/local/Cellar/go@1.13/1.13.9/libexec/src/runtime/panic.go:199\nruntime.sigpanic\n\t/usr/local/Cellar/go@1.13/1.13.9/libexec/src/runtime/signal_unix.go:394\ngithub.com/pingcap/tidb/util/chunk.(*iterator4List).Begin\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/util/chunk/iterator.go:190\ngithub.com/pingcap/tidb/util/chunk.(*iterator4List).Begin\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/util/chunk/iterator.go:190\ngithub.com/pingcap/tidb/executor.(*indexHashJoinInnerWorker).doJoinUnordered\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/executor/index_lookup_hash_join.go:575\ngithub.com/pingcap/tidb/executor.(*indexHashJoinInnerWorker).handleTask\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/executor/index_lookup_hash_join.go:567\ngithub.com/pingcap/tidb/executor.(*indexHashJoinInnerWorker).run\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/executor/index_lookup_hash_join.go:459\ngithub.com/pingcap/tidb/executor.(*IndexNestedLoopHashJoin).startWorkers.func2\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/executor/index_lookup_hash_join.go:185\ngithub.com/pingcap/tidb/util.WithRecovery\n\t/Users/sunrunaway/Code/gopath/src/github.com/pingcap/tidb/util/misc.go:96"]
```

### 4. What is your TiDB version? (Required)

v4.0.4, master

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve?

Issue Number: close #19233  

Problem Summary:

1. Index join does not handle ErrTruncated if join key is ENUM or SET
2. Index Hash Join & Index Lookup Join do not handle error well if some unexpected error happens.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
1. Treat `ErrTruncated` when join key is ENUM or SET
2. Handle `ctx.Err()` in `fetchInnerResults`
3. Also fix some `ctx.Err()` in other places.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- fix index join error when join key is ENUM or SET

### CODE REVIEW

File: executor/index_lookup_merge_join.go [hunk 1]
```diff
@@ -670,6 +671,9 @@ func (imw *innerMergeWorker) constructDatumLookupKey(task *lookUpMergeJoinTask,
 			if terror.ErrorEqual(err, types.ErrOverflow) {
 				return nil, nil
 			}
+			if terror.ErrorEqual(err, types.ErrTruncated) && (innerColType.Tp == mysql.TypeSet || innerColType.Tp == mysql.TypeEnum) {
```
Comment (lzmhhh123) [inline executor/index_lookup_merge_join.go:674]: Does `TypeJson` has the same problem?

Comment (SunRunAway) [reply executor/index_lookup_merge_join.go:674]:
```mysql
mysql> CREATE TABLE `p` (type JSON, key(type));
ERROR 3152 (42000): JSON column 'type' cannot be used in key specification.
```

File: executor/index_lookup_join.go [hunk 2]
```diff
@@ -556,6 +556,9 @@ func (iw *innerWorker) constructDatumLookupKey(task *lookUpJoinTask, chkIdx, row
 			if terror.ErrorEqual(err, types.ErrOverflow) {
 				return nil, nil
 			}
+			if terror.ErrorEqual(err, types.ErrTruncated) && (innerColType.Tp == mysql.TypeSet || innerColType.Tp == mysql.TypeEnum) {
```
Comment (fzhedu) [inline executor/index_lookup_join.go:559]: would `innerColType ` be `outerColType `?

File: executor/index_lookup_join_test.go [hunk 3]
```diff
@@ -212,3 +213,41 @@ func (s *testSuite5) TestIssue16887(c *C) {
...
 }
+
+func (s *testSuite5) TestIndexJoinEnumSetIssue19233(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("drop table if exists i;")
+	tk.MustExec("drop table if exists p1;")
+	tk.MustExec("drop table if exists p2;")
+	tk.MustExec(`CREATE TABLE p1 (type enum('HOST_PORT') NOT NULL, UNIQUE KEY (type)) ;`)
```
Comment (fzhedu) [inline executor/index_lookup_join_test.go:224]: would the error occur if enum == set ? or decimal, json some types?

File: executor/index_lookup_join.go (hunk 2 above)
Comment (SunRunAway) [reply executor/index_lookup_join.go:559]: Only the inner side (index read side) will cause this problem.

File: executor/index_lookup_join_test.go (hunk 3 above)
Comment (SunRunAway) [reply executor/index_lookup_join_test.go:224]: Thanks, I found a new bug, https://github.com/pingcap/tidb/issues/19410

**Pull request merged**